File: tools/loop/lib/state.py
"""
Core loop state management and execution.

This is the heart of Daedalos - the loop execution engine that implements
the Ralph Wiggum technique: iterate until done.

The Loop class orchestrates:
1. Checkpoint creation at each iteration
2. Agent execution with context
3. Promise verification
4. State persistence for pause/resume
"""

import json
import os
import sys
import base64
import time
from pathlib import Path
from datetime import datetime
from dataclasses import dataclass, asdict, field
from typing import Optional, Callable, List, Any
from enum import Enum
import subprocess

from .checkpoint import CheckpointBackend, get_backend
from .agent import AgentAdapter, get_agent, detect_agent
from .promise import verify_promise, verify_promise_detailed, PromiseResult


class LoopStatus(Enum):
    """Status of a loop execution."""
    PENDING = "pending"
    RUNNING = "running"
    PAUSED = "paused"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


@dataclass
class LoopIteration:
    """Record of a single loop iteration."""
    number: int
    started_at: str
    completed_at: Optional[str]
    checkpoint_id: str
    promise_result: Optional[bool]
    promise_output: str
    agent_output: str
    agent_error: str
    changes_summary: str
    duration_ms: int


@dataclass
class LoopState:
    """
    Persistent state for a loop.

    This state is saved to disk after each iteration, enabling:
    - Pause and resume
    - Progress inspection
    - Post-mortem analysis
    """
    id: str
    prompt: str
    promise_cmd: str
    status: LoopStatus
    working_dir: str
    agent_name: str
    max_iterations: int
    current_iteration: int
    iterations: List[LoopIteration]
    created_at: str
    updated_at: str
    initial_checkpoint: str
    injected_context: List[str]
    template: Optional[str] = None
    error_message: Optional[str] = None

    def to_dict(self) -> dict:
        """Convert state to dictionary for JSON serialization."""
        d = asdict(self)
        d["status"] = self.status.value
        return d

    @classmethod
    def from_dict(cls, d: dict) -> "LoopState":
        """Create state from dictionary."""
        d["status"] = LoopStatus(d["status"])
        d["iterations"] = [
            LoopIteration(**i) if isinstance(i, dict) else i
            for i in d.get("iterations", [])
        ]
        return cls(**d)

    def save(self, state_dir: Path):
        """Save state to disk."""
        state_dir.mkdir(parents=True, exist_ok=True)
        path = state_dir / f"{self.id}.json"
        with open(path, "w") as f:
            json.dump(self.to_dict(), f, indent=2)

    @classmethod
    def load(cls, state_dir: Path, loop_id: str) -> "LoopState":
        """Load state from disk."""
        path = state_dir / f"{loop_id}.json"
        with open(path) as f:
            return cls.from_dict(json.load(f))

    @classmethod
    def list_all(cls, state_dir: Path) -> List["LoopState"]:
        """List all saved loop states."""
        states = []
        if state_dir.exists():
            for path in state_dir.glob("*.json"):
                try:
                    states.append(cls.load(state_dir, path.stem))
                except (json.JSONDecodeError, KeyError, TypeError):
                    continue
        return states


class Loop:
    """
    Main loop execution engine.

    This implements the core iteration primitive of Daedalos.

    Usage:
        loop = Loop(
            prompt="fix the failing tests",
            promise_cmd="npm test",
            working_dir=Path("."),
            agent=get_agent("claude"),
            checkpoint=get_backend(Path("."))
        )
        success = loop.run()
    """

    def __init__(
        self,
        prompt: str,
        promise_cmd: str,
        working_dir: Path,
        agent: AgentAdapter,
        checkpoint: CheckpointBackend,
        max_iterations: int = 10,
        timeout: int = 300,
        on_iteration: Optional[Callable[[LoopIteration], None]] = None,
        on_status_change: Optional[Callable[[LoopStatus], None]] = None,
        state_dir: Optional[Path] = None,
        loop_id: Optional[str] = None,
        template: Optional[str] = None
    ):
        """
        Initialize a new loop.

        Args:
            prompt: Natural language task description
            promise_cmd: Shell command that returns 0 when done
            working_dir: Directory for the agent to work in
            agent: Agent adapter to use
            checkpoint: Checkpoint backend to use
            max_iterations: Maximum iterations before giving up
            timeout: Per-iteration timeout in seconds
            on_iteration: Callback after each iteration
            on_status_change: Callback on status changes
            state_dir: Directory for state persistence
            loop_id: Optional ID (auto-generated if not provided)
            template: Name of template being used (for tracking)
        """
        self.prompt = prompt
        self.promise_cmd = promise_cmd
        self.working_dir = Path(working_dir).resolve()
        self.agent = agent
        self.checkpoint = checkpoint
        self.max_iterations = max_iterations
        self.timeout = timeout
        self.on_iteration = on_iteration
        self.on_status_change = on_status_change
        self.state_dir = state_dir or (
            Path.home() / ".local/share/daedalos/loop/states"
        )
        self.state_dir.mkdir(parents=True, exist_ok=True)

        # Initialize state
        self.state = LoopState(
            id=loop_id or self._generate_id(),
            prompt=prompt,
            promise_cmd=promise_cmd,
            status=LoopStatus.PENDING,
            working_dir=str(self.working_dir),
            agent_name=agent.name,
            max_iterations=max_iterations,
            current_iteration=0,
            iterations=[],
            created_at=datetime.now().isoformat(),
            updated_at=datetime.now().isoformat(),
            initial_checkpoint="",
            injected_context=[],
            template=template
        )

    def _generate_id(self) -> str:
        """Generate a unique loop ID."""
        return base64.b64encode(os.urandom(6)).decode().replace('+', '-').replace('/', '_')

    def _set_status(self, status: LoopStatus):
        """Update loop status and trigger callback."""
        self.state.status = status
        self.state.updated_at = datetime.now().isoformat()
        self.state.save(self.state_dir)
        if self.on_status_change:
            self.on_status_change(status)

    def _build_prompt(self, iteration: int) -> str:
        """
        Build the full prompt for an iteration.

        Includes:
        - Task description
        - Iteration number
        - Promise command
        - Injected context
        - Previous iteration feedback
        """
        parts = []

        # Header
        parts.append("=" * 60)
        parts.append(f"LOOP ITERATION {iteration}/{self.max_iterations}")
        parts.append("=" * 60)

        # Task
        parts.append(f"\nTASK:\n{self.prompt}")

        # Promise
        parts.append(f"\nSUCCESS CONDITION:")
        parts.append(f"The following command must exit with code 0:")
        parts.append(f"  {self.promise_cmd}")

        # Injected context
        if self.state.injected_context:
            parts.append("\nADDITIONAL CONTEXT:")
            for ctx in self.state.injected_context:
                parts.append(f"- {ctx}")

        # Previous iteration feedback
        if iteration > 1 and self.state.iterations:
            last = self.state.iterations[-1]
            parts.append(f"\nPREVIOUS ITERATION ({iteration - 1}) RESULT:")

            if last.promise_result:
                parts.append("  Status: PASSED")
            else:
                parts.append("  Status: FAILED")
                if last.promise_output:
                    parts.append("  Output:")
                    for line in last.promise_output.split('\n')[:20]:
                        parts.append(f"    {line}")

            parts.append("\nAnalyze what went wrong and try a different approach.")

        # Instructions
        parts.append("\n" + "=" * 60)
        parts.append("INSTRUCTIONS:")
        parts.append("Make changes to the codebase to satisfy the success condition.")
        parts.append("Focus on the specific task. Make minimal, targeted changes.")
        parts.append("=" * 60)

        return "\n".join(parts)

    def _get_changes_summary(self) -> str:
        """Get a summary of file changes since last checkpoint."""
        try:
            result = subprocess.run(
                ["git", "diff", "--stat", "HEAD"],
                cwd=self.working_dir,
                capture_output=True,
                text=True,
                timeout=10
            )
            if result.stdout.strip():
                return result.stdout.strip()

            # Try unstaged changes
            result = subprocess.run(
                ["git", "diff", "--stat"],
                cwd=self.working_dir,
                capture_output=True,
                text=True,
                timeout=10
            )
            return result.stdout.strip() if result.stdout else "No changes detected"
        except (subprocess.TimeoutExpired, FileNotFoundError):
            return "Unable to detect changes"

    def run(self) -> bool:
        """
        Execute the loop until promise is met or max iterations reached.

        Returns:
            True if promise was met, False otherwise
        """
        # Check if promise is already met
        if verify_promise(self.promise_cmd, self.working_dir):
            self._set_status(LoopStatus.COMPLETED)
            return True

        # Create initial checkpoint
        try:
            self.state.initial_checkpoint = self.checkpoint.create(
                f"{self.state.id}_initial",
                self.working_dir
            )
        except Exception as e:
            self.state.error_message = f"Failed to create initial checkpoint: {e}"
            self._set_status(LoopStatus.FAILED)
            return False

        self._set_status(LoopStatus.RUNNING)

        while self.state.current_iteration < self.max_iterations:
            # Check for pause/cancel
            if self.state.status == LoopStatus.PAUSED:
                time.sleep(1)
                # Reload state in case it changed
                try:
                    self.state = LoopState.load(self.state_dir, self.state.id)
                except:
                    pass
                continue

            if self.state.status == LoopStatus.CANCELLED:
                return False

            # Run iteration
            success = self._run_iteration()
            if success:
                self._set_status(LoopStatus.COMPLETED)
                return True

        # Max iterations reached
        self.state.error_message = f"Max iterations ({self.max_iterations}) reached without meeting promise"
        self._set_status(LoopStatus.FAILED)
        return False

    def _run_iteration(self) -> bool:
        """
        Execute a single loop iteration.

        Returns:
            True if promise was met, False otherwise
        """
        self.state.current_iteration += 1
        iteration_num = self.state.current_iteration
        start_time = time.monotonic()

        # Create checkpoint for this iteration
        try:
            checkpoint_id = self.checkpoint.create(
                f"{self.state.id}_iter{iteration_num}",
                self.working_dir
            )
        except Exception as e:
            checkpoint_id = f"failed_{iteration_num}"

        # Build iteration record
        iteration = LoopIteration(
            number=iteration_num,
            started_at=datetime.now().isoformat(),
            completed_at=None,
            checkpoint_id=checkpoint_id,
            promise_result=None,
            promise_output="",
            agent_output="",
            agent_error="",
            changes_summary="",
            duration_ms=0
        )

        # Build and run agent
        prompt = self._build_prompt(iteration_num)
        agent_result = self.agent.run(
            prompt,
            self.working_dir,
            timeout=self.timeout
        )

        iteration.agent_output = agent_result.output
        iteration.agent_error = agent_result.error

        # Check the promise
        promise_result = verify_promise_detailed(
            self.promise_cmd,
            self.working_dir
        )

        iteration.promise_result = promise_result.success
        iteration.promise_output = (
            promise_result.stdout +
            ("\n" + promise_result.stderr if promise_result.stderr else "")
        )

        # Finalize iteration
        iteration.completed_at = datetime.now().isoformat()
        iteration.changes_summary = self._get_changes_summary()
        iteration.duration_ms = int((time.monotonic() - start_time) * 1000)

        self.state.iterations.append(iteration)
        self.state.updated_at = datetime.now().isoformat()
        self.state.save(self.state_dir)

        # Trigger callback
        if self.on_iteration:
            self.on_iteration(iteration)

        return promise_result.success

    def pause(self):
        """Pause the loop after current iteration."""
        self._set_status(LoopStatus.PAUSED)

    def resume(self):
        """Resume a paused loop."""
        if self.state.status == LoopStatus.PAUSED:
            self._set_status(LoopStatus.RUNNING)

    def cancel(self):
        """Cancel the loop."""
        self._set_status(LoopStatus.CANCELLED)

    def inject_context(self, context: str):
        """Inject additional context for the next iteration."""
        self.state.injected_context.append(context)
        self.state.save(self.state_dir)

    def rollback(self, checkpoint_id: str) -> bool:
        """Rollback to a previous checkpoint."""
        return self.checkpoint.restore(checkpoint_id, self.working_dir)

    def rollback_to_initial(self) -> bool:
        """Rollback to the initial state before the loop started."""
        if self.state.initial_checkpoint:
            return self.rollback(self.state.initial_checkpoint)
        return False

    @classmethod
    def resume_from_state(
        cls,
        state: LoopState,
        agent: Optional[AgentAdapter] = None,
        checkpoint: Optional[CheckpointBackend] = None,
        **kwargs
    ) -> "Loop":
        """
        Resume a loop from saved state.

        Args:
            state: Previously saved loop state
            agent: Agent to use (auto-detects if not provided)
            checkpoint: Checkpoint backend (auto-detects if not provided)
            **kwargs: Additional arguments to override

        Returns:
            Loop instance ready to resume
        """
        working_dir = Path(state.working_dir)

        if agent is None:
            agent = get_agent(state.agent_name)
        if checkpoint is None:
            checkpoint = get_backend(working_dir)

        loop = cls(
            prompt=state.prompt,
            promise_cmd=state.promise_cmd,
            working_dir=working_dir,
            agent=agent,
            checkpoint=checkpoint,
            max_iterations=state.max_iterations,
            loop_id=state.id,
            template=state.template,
            **kwargs
        )

        # Restore state
        loop.state = state

        return loop


def get_loop(loop_id: str, state_dir: Optional[Path] = None) -> Optional[LoopState]:
    """
    Get loop state by ID.

    Args:
        loop_id: The loop identifier
        state_dir: Directory containing state files

    Returns:
        LoopState if found, None otherwise
    """
    state_dir = state_dir or Path.home() / ".local/share/daedalos/loop/states"
    try:
        return LoopState.load(state_dir, loop_id)
    except FileNotFoundError:
        return None


def list_loops(
    state_dir: Optional[Path] = None,
    status: Optional[LoopStatus] = None
) -> List[LoopState]:
    """
    List all loops, optionally filtered by status.

    Args:
        state_dir: Directory containing state files
        status: Filter to only this status

    Returns:
        List of LoopState objects
    """
    state_dir = state_dir or Path.home() / ".local/share/daedalos/loop/states"
    states = LoopState.list_all(state_dir)

    if status:
        states = [s for s in states if s.status == status]

    return sorted(states, key=lambda s: s.updated_at, reverse=True)

File: crates/loop/src/state.rs
//! Loop state management
//!
//! Handles persistence and execution of loops.

use anyhow::{Context, Result};
use chrono::{DateTime, Utc};
use daedalos_core::Paths;
use serde::{Deserialize, Serialize};
use std::path::PathBuf;
use std::time::Instant;

use crate::checkpoint::CheckpointBackend;
use crate::promise::verify_promise_detailed;

/// Status of a loop
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum LoopStatus {
    Pending,
    Running,
    Paused,
    Completed,
    Failed,
    Cancelled,
}

impl LoopStatus {
    pub fn as_str(&self) -> &'static str {
        match self {
            LoopStatus::Pending => "pending",
            LoopStatus::Running => "running",
            LoopStatus::Paused => "paused",
            LoopStatus::Completed => "completed",
            LoopStatus::Failed => "failed",
            LoopStatus::Cancelled => "cancelled",
        }
    }
}

/// Record of a single loop iteration
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LoopIteration {
    pub number: u32,
    pub started_at: DateTime<Utc>,
    pub completed_at: Option<DateTime<Utc>>,
    pub checkpoint_id: String,
    pub promise_result: bool,
    pub promise_output: String,
    pub promise_exit_code: i32,
    pub changes_summary: String,
    pub duration_ms: u64,
}

/// Persistent state for a loop
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LoopState {
    pub id: String,
    pub prompt: String,
    pub promise_cmd: String,
    pub status: LoopStatus,
    pub working_dir: PathBuf,
    pub max_iterations: u32,
    pub current_iteration: u32,
    pub iterations: Vec<LoopIteration>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub initial_checkpoint: Option<String>,
    pub injected_context: Vec<String>,
    pub checkpoint_strategy: String,
    pub timeout: u64,
    pub error_message: Option<String>,
}

impl LoopState {
    /// Create a new loop state
    pub fn new(config: &LoopConfig) -> Self {
        let id = generate_loop_id();
        let now = Utc::now();

        Self {
            id,
            prompt: config.prompt.clone(),
            promise_cmd: config.promise_cmd.clone(),
            status: LoopStatus::Pending,
            working_dir: config.working_dir.clone(),
            max_iterations: config.max_iterations,
            current_iteration: 0,
            iterations: Vec::new(),
            created_at: now,
            updated_at: now,
            initial_checkpoint: None,
            injected_context: Vec::new(),
            checkpoint_strategy: config.checkpoint_strategy.clone(),
            timeout: config.timeout,
            error_message: None,
        }
    }

    /// Get the state directory
    fn state_dir() -> PathBuf {
        let paths = Paths::new();
        paths.state("loop").join("states")
    }

    /// Save state to disk
    pub fn save(&self) -> Result<()> {
        let state_dir = Self::state_dir();
        std::fs::create_dir_all(&state_dir).context("Failed to create state directory")?;

        let path = state_dir.join(format!("{}.json", self.id));
        let content = serde_json::to_string_pretty(self)?;
        std::fs::write(&path, content).context("Failed to write state file")?;

        Ok(())
    }

    /// Load state from disk
    pub fn load(loop_id: &str) -> Result<Self> {
        let state_dir = Self::state_dir();
        let path = state_dir.join(format!("{}.json", loop_id));

        if !path.exists() {
            anyhow::bail!("Loop not found: {}", loop_id);
        }

        let content = std::fs::read_to_string(&path).context("Failed to read state file")?;
        let state: Self = serde_json::from_str(&content).context("Failed to parse state file")?;

        Ok(state)
    }

    /// List all loop states
    pub fn list_all() -> Result<Vec<Self>> {
        let state_dir = Self::state_dir();
        let mut states = Vec::new();

        if !state_dir.exists() {
            return Ok(states);
        }

        for entry in std::fs::read_dir(&state_dir)? {
            let entry = entry?;
            let path = entry.path();

            if path.extension().map(|e| e == "json").unwrap_or(false) {
                match std::fs::read_to_string(&path) {
                    Ok(content) => {
                        if let Ok(state) = serde_json::from_str::<Self>(&content) {
                            states.push(state);
                        }
                    }
                    Err(_) => continue,
                }
            }
        }

        // Sort by updated_at descending
        states.sort_by(|a, b| b.updated_at.cmp(&a.updated_at));

        Ok(states)
    }
}

/// Configuration for starting a loop
pub struct LoopConfig {
    pub prompt: String,
    pub promise_cmd: String,
    pub working_dir: PathBuf,
    pub max_iterations: u32,
    pub timeout: u64,
    pub checkpoint_strategy: String,
}

/// Main loop execution engine
pub struct Loop {
    state: LoopState,
    checkpoint: Box<dyn CheckpointBackend>,
}

impl Loop {
    /// Create a new loop
    pub fn new(config: LoopConfig, checkpoint: Box<dyn CheckpointBackend>) -> Result<Self> {
        let state = LoopState::new(&config);

        Ok(Self { state, checkpoint })
    }

    /// Get the current state
    pub fn state(&self) -> &LoopState {
        &self.state
    }

    /// Get checkpoint backend name
    pub fn checkpoint_name(&self) -> &str {
        self.checkpoint.name()
    }

    /// Run the loop until promise is met or max iterations reached
    pub async fn run(&mut self) -> Result<bool> {
        // Create initial checkpoint
        match self.checkpoint.create(
            &format!("{}_initial", self.state.id),
            &self.state.working_dir,
        ) {
            Ok(checkpoint_id) => {
                self.state.initial_checkpoint = Some(checkpoint_id);
            }
            Err(e) => {
                self.state.error_message = Some(format!("Failed to create initial checkpoint: {}", e));
                self.state.status = LoopStatus::Failed;
                self.state.save()?;
                return Ok(false);
            }
        }

        self.state.status = LoopStatus::Running;
        self.state.save()?;

        while self.state.current_iteration < self.state.max_iterations {
            // Check for pause/cancel
            if self.state.status == LoopStatus::Cancelled {
                return Ok(false);
            }

            // Run iteration
            let success = self.run_iteration().await?;
            if success {
                self.state.status = LoopStatus::Completed;
                self.state.save()?;
                return Ok(true);
            }
        }

        // Max iterations reached
        self.state.error_message = Some(format!(
            "Max iterations ({}) reached without meeting promise",
            self.state.max_iterations
        ));
        self.state.status = LoopStatus::Failed;
        self.state.save()?;

        Ok(false)
    }

    /// Run a single iteration
    async fn run_iteration(&mut self) -> Result<bool> {
        self.state.current_iteration += 1;
        let iteration_num = self.state.current_iteration;
        let start_time = Instant::now();

        // Create checkpoint for this iteration
        let checkpoint_id = match self.checkpoint.create(
            &format!("{}_iter{}", self.state.id, iteration_num),
            &self.state.working_dir,
        ) {
            Ok(id) => id,
            Err(_) => format!("failed_{}", iteration_num),
        };

        // Print iteration info
        println!("{}", "=".repeat(60));
        println!("ITERATION {}/{}", iteration_num, self.state.max_iterations);
        println!("{}", "=".repeat(60));
        println!();

        // Send prompt to oracle (the LLM primitive)
        let prompt = self.build_prompt(iteration_num);
        println!("Sending prompt to oracle...");
        println!();

        if let Err(e) = self.invoke_oracle(&prompt).await {
            eprintln!("Warning: oracle invocation failed: {}", e);
            // Continue anyway - user might be running an external agent
        }

        println!();
        println!("Running verification: {}", self.state.promise_cmd);

        // Verify the promise
        let promise_result =
            verify_promise_detailed(&self.state.promise_cmd, &self.state.working_dir, self.state.timeout)
                .await?;

        let duration = start_time.elapsed();

        // Get changes summary
        let changes_summary = get_changes_summary(&self.state.working_dir).await;

        // Create iteration record
        let iteration = LoopIteration {
            number: iteration_num,
            started_at: Utc::now() - chrono::Duration::milliseconds(duration.as_millis() as i64),
            completed_at: Some(Utc::now()),
            checkpoint_id,
            promise_result: promise_result.success,
            promise_output: format!("{}\n{}", promise_result.stdout, promise_result.stderr),
            promise_exit_code: promise_result.exit_code,
            changes_summary,
            duration_ms: duration.as_millis() as u64,
        };

        // Print result
        let status_str = if promise_result.success { "PASS" } else { "FAIL" };
        println!();
        println!(
            "[Iteration {}] {} ({}ms)",
            iteration_num,
            status_str,
            iteration.duration_ms
        );

        if !promise_result.success && !promise_result.stdout.is_empty() {
            println!();
            // Show first few lines of output
            for line in promise_result.stdout.lines().take(10) {
                println!("  {}", line);
            }
        }

        self.state.iterations.push(iteration);
        self.state.updated_at = Utc::now();
        self.state.save()?;

        Ok(promise_result.success)
    }

    /// Build the full prompt for an iteration
    fn build_prompt(&self, iteration: u32) -> String {
        let mut parts = Vec::new();

        parts.push("=".repeat(60));
        parts.push(format!(
            "LOOP ITERATION {}/{}",
            iteration, self.state.max_iterations
        ));
        parts.push("=".repeat(60));

        parts.push(format!("\nTASK:\n{}", self.state.prompt));

        parts.push("\nSUCCESS CONDITION:".to_string());
        parts.push("The following command must exit with code 0:".to_string());
        parts.push(format!("  {}", self.state.promise_cmd));

        // Injected context
        if !self.state.injected_context.is_empty() {
            parts.push("\nADDITIONAL CONTEXT:".to_string());
            for ctx in &self.state.injected_context {
                parts.push(format!("- {}", ctx));
            }
        }

        // Previous iteration feedback
        if iteration > 1 {
            if let Some(last) = self.state.iterations.last() {
                parts.push(format!("\nPREVIOUS ITERATION ({}) RESULT:", iteration - 1));

                if last.promise_result {
                    parts.push("  Status: PASSED".to_string());
                } else {
                    parts.push("  Status: FAILED".to_string());
                    if !last.promise_output.is_empty() {
                        parts.push("  Output:".to_string());
                        for line in last.promise_output.lines().take(20) {
                            parts.push(format!("    {}", line));
                        }
                    }
                }

                parts.push("\nAnalyze what went wrong and try a different approach.".to_string());
            }
        }

        parts.push("\n".to_string() + &"=".repeat(60));
        parts.push("INSTRUCTIONS:".to_string());
        parts.push("Make changes to the codebase to satisfy the success condition.".to_string());
        parts.push("Focus on the specific task. Make minimal, targeted changes.".to_string());
        parts.push("=".repeat(60));

        parts.join("\n")
    }

    /// Inject additional context for the next iteration
    pub fn inject_context(&mut self, context: String) -> Result<()> {
        self.state.injected_context.push(context);
        self.state.save()?;
        Ok(())
    }

    /// Cancel the loop
    pub fn cancel(&mut self) -> Result<()> {
        self.state.status = LoopStatus::Cancelled;
        self.state.save()?;
        Ok(())
    }
}

/// Generate a unique loop ID
fn generate_loop_id() -> String {
    let bytes: [u8; 6] = rand_bytes();
    base64::Engine::encode(&base64::engine::general_purpose::URL_SAFE_NO_PAD, bytes)
}

/// Generate random bytes (simple implementation)
fn rand_bytes() -> [u8; 6] {
    use std::time::{SystemTime, UNIX_EPOCH};
    let nanos = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .subsec_nanos();
    let mut bytes = [0u8; 6];
    bytes[0..4].copy_from_slice(&nanos.to_le_bytes());
    bytes[4] = (nanos >> 8) as u8;
    bytes[5] = (nanos >> 16) as u8;
    bytes
}

/// Get a summary of file changes
async fn get_changes_summary(working_dir: &PathBuf) -> String {
    let output = tokio::process::Command::new("git")
        .args(["diff", "--stat", "HEAD"])
        .current_dir(working_dir)
        .output()
        .await;

    match output {
        Ok(out) if out.status.success() => {
            let stdout = String::from_utf8_lossy(&out.stdout).to_string();
            if !stdout.trim().is_empty() {
                return stdout;
            }

            // Try unstaged changes
            let unstaged = tokio::process::Command::new("git")
                .args(["diff", "--stat"])
                .current_dir(working_dir)
                .output()
                .await;

            match unstaged {
                Ok(out) => String::from_utf8_lossy(&out.stdout).to_string(),
                Err(_) => "Unable to detect changes".to_string(),
            }
        }
        _ => "Unable to detect changes".to_string(),
    }
}

/// List all loops
pub fn list_loops() -> Result<Vec<LoopState>> {
    LoopState::list_all()
}
